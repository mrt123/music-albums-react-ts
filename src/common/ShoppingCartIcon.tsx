import { Link } from "react-router-dom";

const linkStyle = {
  textDecoration: "none",
  color: "white",
};

const getNumberOfItemsStyle = (backgroundColor: string | undefined) => {
  return {
    position: "absolute" as "absolute",
    top: "17px",
    left: "18px",
    textAlign: "center" as "center",
    background: backgroundColor ? backgroundColor : "yellow",
    borderRadius: "28%",
    border: "2px solid grey",
    height: "28px",
    padding: "0 3px",
    fontSize: "25px",
    color: "black",
  };
};

interface ShoppingCartIconProps {
  numberOfItems: number;
  numberBackgroundColor?: string;
}

const ShoppingCartIcon: React.FC<ShoppingCartIconProps> = ({
  numberOfItems,
  numberBackgroundColor,
}) => {
  const numberIcon =
    numberOfItems > 0 ? (
      <span style={getNumberOfItemsStyle(numberBackgroundColor)}>
        {numberOfItems}
      </span>
    ) : null;

  return (
    <div style={{ position: "relative" }}>
      <Link to="/shopping-cart" style={linkStyle}>
        ðŸ›’
        {numberIcon}
      </Link>
    </div>
  );
};

export default ShoppingCartIcon;
