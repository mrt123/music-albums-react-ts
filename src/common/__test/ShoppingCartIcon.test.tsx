import { render, RenderOptions } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import ShoppingCartIcon from "../ShoppingCartIcon";

/**
 * render(<ShoppingCartIcon numberOfItems={0} />, {
      wrapper: BrowserRouter,
    });

    Powysze powtarza się w naszym tescie ... czy mozemy to jakoś usprawnić.

    Versja na 5: zapewnimy jakąkolwiek formę anstrakcji, tak zebyśmy mogli ten kod re-uzyc
    Versja na 6: napiszemy naszego własnego renderera, który jest oparty na 
    (i 100% kompatybliny z ) render z bilbioteki react-router-dom. 
    Podpowiedźio pytanie: Czym jest renderer? Odp: Funkcją ;-) 
 */

const renderShoppingCartIcon = (
  numberOfItems: number,
  options?: RenderOptions
) => {
  return render(<ShoppingCartIcon numberOfItems={numberOfItems} />, {
    wrapper: BrowserRouter,
    ...options,
  });
};

describe("ShoppingCartIcon.tsx", () => {
  it("matchesSnapshot", () => {
    const { container } = renderShoppingCartIcon(99);
    expect(container).toMatchSnapshot();
  });

  it("zero is not rendered numberOfItems is 0", () => {
    const { queryByTestId } = renderShoppingCartIcon(0);

    const numbnerIcon = queryByTestId("number-icon");
    expect(numbnerIcon).toBeNull();
  });

  it("5 is rendered when numberOfItems is 5", async () => {
    const { findByTestId } = renderShoppingCartIcon(5);

    const numbnerIcon = await findByTestId("number-icon");
    expect(numbnerIcon).toHaveTextContent("5");
  });
});
