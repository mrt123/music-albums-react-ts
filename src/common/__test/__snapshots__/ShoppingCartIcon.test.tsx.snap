// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCartIcon.tsx matchesSnapshot 1`] = `
<div>
  <div
    data-testid="shopping-cart-icon"
    style="position: relative;"
  >
    <a
      href="/shopping-cart"
      style="text-decoration: none; color: white;"
    >
      ðŸ›’
      <span
        data-testid="number-icon"
        style="position: absolute; top: 17px; left: 18px; text-align: center; background: yellow; border-radius: 28%; border: 2px solid grey; height: 28px; padding: 0px 3px; font-size: 25px; color: black;"
      >
        99
      </span>
    </a>
  </div>
</div>
`;
